Add Cloud Translate Services

SarahZLipeles opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #N/A
2. This PR does the following: Creates Cloud Translate Services using the Google Cloud Translate API.
Creates translate_text function
Creates Cloud Translate Emulator for testing
Adds Google Cloud Translate to requirements

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
Manual testing:
```python
>>> execfile("./core/platform/cloud_translate/cloud_translate_services.py")
>>> print(translate_text("hello world", "en", "ar"))
مرحبا بالعالم
>>> print(translate_text("hello world", "en", "es"))
Hola Mundo
```

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Commit:
add cloud translate services and emulator

Commit:
add cloud translate services and emulator tests

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

Commit:
add google cloud translate api to requirements

Commit:
fix duplicate test names

Commit:
fix presubmit errors

SarahZLipeles commented:
@seanlip @sagangwee @brianrodri @U8NWXD @vojtechjelinek PTAL

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,69 @@
+            response = self.expected_responses.get(key)
+        return response
+
+    def add_expected_response(self, inputs, response):
```
Separate out `inputs` into individual parameters. In general, try not to pass around things like tuples and dicts. If you need to group the three things together then create a Python domain object for that.

SarahZLipeles replied:
Thanks, I separated these out.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,69 @@
+        """Initializes the emulator with no expected responses."""
+        self.expected_responses = {}
+
+    def translate_text(self, text, source_language, target_language):
```
"language" is not the same as "language code" -- please rename accordingly.

SarahZLipeles replied:
Done.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator_test.py:
```diff
@@ -0,0 +1,53 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for cloud_translate_services."""
```
Add a new line above this one.

SarahZLipeles replied:
Done.

TheoLipeles replied:
Done.

seanlip commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,48 @@
+        str. The translated text.
+    """
+    if not utils.is_valid_language_code(source_language):
+        raise ValueError('invalid language code: %s' % source_language)
```
Should these error messages start with a capital letter?

SarahZLipeles replied:
Yes, I changed them to be capitalized properly.

seanlip replied:
They don't look changed to me?

SarahZLipeles replied:
Ah, I had changed them in `cloud_translate_emulator`, but not here.
Done.

seanlip reviewed (changes requested):
Looks pretty good -- thanks @TheoLipeles! Left a few inline comments.

seanlip commented:
Also, one other question. If this PR is merged into develop, and then develop is deployed to a production server (without any other PRs being merged), will the server break? If so, or if there are setup steps that need to be tested first, then we need to do a test run of this in a prod environment before this PR gets merged. Note that you can request that something be tested on prod following [these instructions](https://github.com/oppia/oppia/wiki/Running-jobs-in-production) (they're for jobs, but you can also use it for something like this, just make sure the instructions in the doc you submit with the request are very clear). This will avoid us breaking production by accident.

sagangwee commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,69 @@
+        self.expected_responses = {}
+
+    def translate_text(self, text, source_language, target_language):
+        """Responds with the saved expected response for a given input. If no
```
Nit: Instead of "Responds with" can just say "Returns".

SarahZLipeles replied:
Done.

sagangwee commented on core/platform/cloud_translate/cloud_translate_services_test.py:
```diff
@@ -0,0 +1,59 @@
+    def test_translate_text_with_invalid_source_language_raises_error(self):
+        with self.assertRaisesRegexp(
+            ValueError, 'invalid language code: invalid'):
+
```
Nit: New line not necessary here. Ditto below.

SarahZLipeles replied:
I removed the newline.
I had added it because the indent from the line above makes it hard to tell where the `with` statement ends, is there a better solution for that?

sagangwee commented on core/platform/cloud_translate/cloud_translate_services_test.py:
```diff
@@ -0,0 +1,59 @@
+
+        def mock_translate_call(
+                text, source_language, target_language):
+            self.assertEqual(text, 'hello world')
+            self.assertEqual(source_language, 'en')
+            self.assertEqual(target_language, 'es')
```
Don't think these asserts are necessary as the arguments passed to the mock function isn't what we really care about testing here.

SarahZLipeles replied:
The arguments need to be tested or the linter will throw an unused parameter error. Additionally since some parameters are referenced by name in `cloud_translate_services`, these asserts serve the purpose of guaranteeing that the client API is called correctly.

sagangwee replied:
Ah, I see. If you want to verify that the translate function is called with particular arguments, probably better to use `test_utils.swap_with_checks()` here instead of "hiding" the asserts in the mock function. That way what you're asserting is more explicit. If you're running into the unused param lint issue, I think it's fine to disable that check in this case.

sagangwee reviewed (changes requested):
@TheoLipeles Left a few comments. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

sajalasati commented:
@TheoLipeles Going through the PR thread, I think you need to address the review comments from previous reviews. 

@brianrodri you may un-assign yourself if the PR is not yet ready for you to review. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

Commit:
fix spacing

Commit:
fix parameter naming and convention
replace tuple input with individual inputs
clarify language vs language_code

seanlip commented:
Per discussion with @TheoLipeles, I'm going to close this for now. We can pick it up later when it makes sense to work on it again, but some more work needs to be done here to verify that the proposed integration would work correctly in production.

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

SarahZLipeles commented:
This PR is now up to date and tested on a backup server, [this branch](https://github.com/TheoLipeles/oppia/tree/google-translate-api_test-page) was used for testing.

SarahZLipeles commented:
@sagangwee @seanlip @brianrodri PTAL

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator_test.py:
```diff
@@ -0,0 +1,54 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for cloud_translate_services."""
```
Please be clearer here, these are tests for the cloud translate emulator, not cloud_translate_services.

SarahZLipeles replied:
Fixed.
Thanks for catching this!

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator_test.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class CloudTranslateEmulatorUnitTests(test_utils.TestBase):
+    """Tests for cloud_translate_services."""
```
Ditto, doesn't match the class/module you're testing.

SarahZLipeles replied:
Fixed.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,73 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An emulator that mocks the core.platform.cloud_translate API. This emulator
```
Is this only for the backend test env, or is it for the local dev environment too? Please clarify in the docstring.

More generally, where is this file actually used? I don't see anything else referencing it in this PR.

SarahZLipeles replied:
I added some clarification, the emulator is not currently used, rather, it exists for backend testing and possibly local dev environment.
In the local dev environment, it's much easier to create a free tier GCP and set up authentication (I added a link and instructions for doing so). But in the case that authentication is not an option, the emulator can be used for local testing.

seanlip replied:
Sorry, I still don't understand. In platform/models.py, when calling import_cloud_translate_services(), we only return the non-emulator module, so this is what will be selected by default in the local dev environment. If we want the local dev environment to work, it needs to work "out of the box", just like all our other third-party platform dependencies. Developers who are not working with Cloud Translate should not need to create a free-tier GCP, etc.

You'll probably want the code in platform/models.py to choose the correct implementation based on whether we're in EMULATOR_MODE or not, similar to how import_taskqueue_services works. I suggest changing the naming here to be consistent with that example, too. Note, also, that the different implementations need to expose the same API so that it is possible to swap between them.

SarahZLipeles replied:
I updated this accordingly, let me know if this is what you meant.

seanlip commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,48 @@
+CLIENT = translate.Client()
+
+
+def translate_text(text, source_language, target_language):
```
Is this meant to be used in local dev mode (localhost:8181) as well? If so ... does it actually work when you test manually as well (e.g. by adding some controller that makes use of import_cloud_translate_services() in platform/models)?

SarahZLipeles replied:
See above.
This does not work for manual testing unless the user authenticates using their own GCP service account. Otherwise, the emulator is preferred.

seanlip replied:
Replied above, I think there might still be issues here.

SarahZLipeles replied:
I tested this with no credentials and it does not cause problems.

seanlip reviewed (changes requested):
Thanks @TheoLipeles! Took a full pass, PTAL.

Commit:
Fix capitalization
Add comments clarifying usage instructions
Fix incorrect docstrings in cloud_translate_emulator

SarahZLipeles commented:
@seanlip I responded to your comments, PTAL.

sagangwee commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,74 @@
+
+        Returns:
+            str. The translated text.
+        """
```
Add the possible raised exceptions in the docstring as well.

SarahZLipeles replied:
Added.

sagangwee reviewed (changes requested):
@TheoLipeles Thanks, just a couple comments.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,74 @@
+
+    def translate_text(self, text, source_language_code, target_language_code):
+        """Returns the saved expected response for a given input. If no
+        response exists for the given input, returns an empty string.
```
But the function is returning the response if the input doesn't exist, so which is right?

SarahZLipeles replied:
I updated this comment to mention a default response. Is this still confusing?

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,74 @@
+        if not utils.is_valid_language_code(target_language_code):
+            raise ValueError('Invalid language code: %s' % target_language_code)
+
+        response = ''
+        key = (source_language_code, target_language_code, text)
+        if key in self.expected_responses:
+            response = self.expected_responses.get(key)
+        return response
```
Please just return the values directly, the usage of response was confusing at first because it seemed like an input argument.

```suggestion
        key = (source_language_code, target_language_code, text)
        if key in self.expected_responses:
            return self.expected_responses.get(key)
        return ''
```

SarahZLipeles replied:
This has been updated with a default response and a comment labeling it. Is this more clear?

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,74 @@
+            response: str. The expected response for the given inputs.
+        """
+        inputs = (source_language_code, target_language_code, source_text)
+        self.expected_responses.update({inputs: response})
```
Please just use dict assignment:
```suggestion
        self.expected_responses[inputs] = response
```

SarahZLipeles replied:
Fixed.

brianrodri commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,54 @@
+    Args:
+        text: str. The text to be translated.
+        source_language: str. A valid ISO 639-1 language code.
+        target_language: str. A valid ISO 639-1 language code.
```
Please add a blurb or link explaining what ISO 639-1 is to save readers the time of looking it up themselves.

SarahZLipeles replied:
I added a link.
This has also been updated to reflect the use of a language code whitelist.

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

Commit:
add clarification on ISO 639-1

Commit:
use cloud_translate_emulator in EMULATOR_MODE

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

Commit:
fix linter issues

Commit:
update comment

SarahZLipeles commented:
@seanlip @sagangwee @brianrodri PTAL
I responded to your comments.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,92 @@
+
+# Pre-generated translations for the following phrases:
+# ('hello world', 'CONTINUE', 'Please continue.', 'Correct!')
+PREGENERATED_TRANSLATIONS = {(u'en', u'zh', u'Please continue.'): u'\u8bf7\u7ee7\u7eed\u3002', (u'en', u'es', u'Please continue.'): u'Por favor continua.', (u'en', u'fr', u'CONTINUE'): u'CONTINUEZ', (u'en', u'fr', u'Please continue.'): u'Continuez s&#39;il vous pla\xeet.', (u'en', u'es', u'CONTINUE'): u'SEGUIR', (u'en', u'zh', u'hello world'): u'\u4f60\u597d\u4e16\u754c', (u'en', u'es', u'Correct ...
```
Please break after "{" and put each key-value pair on a separate line. Otherwise, this is really hard to read.

In general, avoid using pylint-disable. If you need to do that, something is wrong.

Also, I suggest having this constant live as a class-level constant inside the TranslateEmulator (possibly as a private member).

SarahZLipeles replied:
Fixed.
I didn't make this private because it's used in testing, do suggest keeping this private?

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,92 @@
+
+        response = (
+            'Default translation. (See core/platform/cloud_translate/'
+            'cloud_translate_emulator.py for details)') # Default response.
```
Just call this default_response, or better still, make it a class-level _DEFAULT_RESPONSE constant.

SarahZLipeles replied:
I added a class-level DEFAULT_RESPONSE constant, it's not private for the same reason as above.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,92 @@
+
+        key = (source_language_code, target_language_code, text)
+        if key in self.expected_responses:
+            response = self.expected_responses[key]
```
Can replace these three lines by: `response = self.expected_responses.get(key, default=self._DEFAULT_RESPONSE)`

SarahZLipeles replied:
Thanks! This is much cleaner.
Done.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,92 @@
+            raise ValueError('Invalid language code: %s' % target_language_code)
+
+        response = (
+            'Default translation. (See core/platform/cloud_translate/'
```
Maybe say "Default translation for emulator mode." so that if someone sees this outside of emulator mode, they know something's gone wrong.

SarahZLipeles replied:
Sounds good.
Done.

seanlip commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,92 @@
+PREGENERATED_TRANSLATIONS = {(u'en', u'zh', u'Please continue.'): u'\u8bf7\u7ee7\u7eed\u3002', (u'en', u'es', u'Please continue.'): u'Por favor continua.', (u'en', u'fr', u'CONTINUE'): u'CONTINUEZ', (u'en', u'fr', u'Please continue.'): u'Continuez s&#39;il vous pla\xeet.', (u'en', u'es', u'CONTINUE'): u'SEGUIR', (u'en', u'zh', u'hello world'): u'\u4f60\u597d\u4e16\u754c', (u'en', u'es', u'Correct ...
+
+
+class TranslateEmulator(python_utils.OBJECT):
```
Perhaps call it CloudTranslateEmulator since that's what you're calling it in the filename and tests as well.

SarahZLipeles replied:
Sounds good.
Done.

seanlip commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,56 @@
+    Returns:
+        str. The translated text.
+    """
+    if source_language not in LANGUAGE_CODE_WHITELIST:
```
Could you use ALLOWLIST instead of WHITELIST? We are trying to follow the allowlist/denylist convention throughout the codebase.

SarahZLipeles replied:
I changed all "whitelist" references to "allowlist".

seanlip commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,56 @@
+CLIENT = translate.Client()
+
+# List of languages with adequate Google Translate accuracy.
+LANGUAGE_CODE_WHITELIST = ('en', 'es', 'fr', 'zh')
```
Please add Portuguese too (pt) -- I checked with Anu and she suggested that as well. Thanks!

SarahZLipeles replied:
Added.

seanlip commented on core/platform/models.py:
```diff
@@ -300,6 +300,20 @@ def import_search_services(cls):
+        Returns:
+            module. The cloud_translate_services module.
+        """
+        if constants.EMULATOR_MODE:
```
Remember my comment above: cloud_translate_emulator and cloud_translate_services should expose the same API. Currently, they do not. The first exposes a TranslateEmulator class and the second exposes a translate_text() function.

Please make these consistent. Probably add a translate_text() top-level function to the first one (I'm assuming the functions in the rest of the codebase are going to be calling only that and not invoking the TranslateEmulator directly.

SarahZLipeles replied:
I changed this to follow the same pattern as cloud_taskqueue_services. It now uses dev_mode_cloud_translate_services.

Commit:
add dev_mode_cloud_translate_services
switch "whitelist" term to "allowlist"
add Portuguese to language allowlist
rename TranslateEmulator to CloudTranslateEmulator
cleanup code

SarahZLipeles commented:
@seanlip I responded to your comments, PTAL.

seanlip commented on core/platform/cloud_translate/cloud_translate_services_test.py:
```diff
@@ -0,0 +1,73 @@
+
+    def test_translate_text_with_invalid_source_language_raises_error(self):
+        with self.assertRaisesRegexp(
+            # Hindi (hi) is not a whitelisted language code.
```
Please change use of "whitelist" to "allowlist" throughout this PR.

SarahZLipeles replied:
Fixed. I did a search through all cloud_translate files.

seanlip reviewed (approved):
Thanks @TheoLipeles, LGTM! Just had one comment.

seanlip commented:
@brianrodri @sagangwee PTAL as codeowners, thanks!

Also, @TheoLipeles, note that a backend test is failing.

Commit:
replace "whitelisted" with "allowlisted"

Commit:
Merge remote-tracking branch 'upstream/develop' into google-translate-api

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,107 @@
+    """
+
+    def __init__(self):
+        """Initializes the emulator with no expected responses."""
```
Seems this is wrong, since we're assigning a non-empty value to self.expected_responses. Please fix the docstring or make the responses actually empty.

SarahZLipeles replied:
Thanks for catching this!
Fixed.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,107 @@
+    must be passed in before using this emulator for testing. See
+    PREGENERATED_TRANSLATIONS above for some prepopulated responses.
+    """
+
+    def __init__(self):
+        """Initializes the emulator with no expected responses."""
+        self.DEFAULT_RESPONSE = (
+            'Default translation for emulator mode. (See core/platform/'
+            'cloud_translate/cloud_translate_emulator.py for details)')
+
+        # Pre-generated translations for the following phrases:
+        # ('hello world', 'CONTINUE', 'Please continue.', 'Correct!')
+        self.PREGENERATED_TRANSLATIONS = {
+            (u'en', u'pt', u'hello world'): u'Ol\xe1 Mundo',
+            (u'en', u'pt', u'CONTINUE'): u'PROSSEGUIR',
+            (u'en', u'es', u'Please continue.'): u'Por favor continua.',
+            (u'en', u'fr', u'CONTINUE'): u'CONTINUEZ',
+            (u'en', u'fr', u'Please continue.'):
+                u'Continuez s&#39;il vous pla\xeet.',
+            (u'en', u'es', u'CONTINUE'): u'SEGUIR',
+            (u'en', u'zh', u'hello world'): u'\u4f60\u597d\u4e16\u754c',
+            (u'en', u'es', u'Correct!'): u'\xa1Correcto!',
+            (u'en', u'zh', u'Correct!'): u'\u6b63\u786e\u7684\uff01',
+            (u'en', u'zh', u'CONTINUE'): u'\u7ee7\u7eed',
+            (u'en', u'zh', u'Please continue.'): u'\u8bf7\u7ee7\u7eed\u3002',
+            (u'en', u'fr', u'Correct!'): u'Correct!',
+            (u'en', u'pt', u'Correct!'): u'Correto!',
+            (u'en', u'es', u'hello world'): u'Hola Mundo',
+            (u'en', u'pt', u'Please continue.'): u'Por favor continue.',
+            (u'en', u'fr', u'hello world'): u'Bonjour le monde'
+            }
+        self.expected_responses = self.PREGENERATED_TRANSLATIONS
```
Please move class constants into the class body:

```suggestion
    DEFAULT_RESPONSE = (
        'Default translation for emulator mode. (See core/platform/'
        'cloud_translate/cloud_translate_emulator.py for details)')

    # Pre-generated translations for the following phrases:
    # ('hello world', 'CONTINUE', 'Please continue.', 'Correct!')
    PREGENERATED_TRANSLATIONS = {
        (u'en', u'pt', u'hello world'): u'Ol\xe1 Mundo',
        (u'en', u'pt', u'CONTINUE'): u'PROSSEGUIR',
        (u'en', u'es', u'Please continue.'): u'Por favor continua.',
        (u'en', u'fr', u'CONTINUE'): u'CONTINUEZ',
        (u'en', u'fr', u'Please continue.'):
            u'Continuez s&#39;il vous pla\xeet.',
        (u'en', u'es', u'CONTINUE'): u'SEGUIR',
        (u'en', u'zh', u'hello world'): u'\u4f60\u597d\u4e16\u754c',
        (u'en', u'es', u'Correct!'): u'\xa1Correcto!',
        (u'en', u'zh', u'Correct!'): u'\u6b63\u786e\u7684\uff01',
        (u'en', u'zh', u'CONTINUE'): u'\u7ee7\u7eed',
        (u'en', u'zh', u'Please continue.'): u'\u8bf7\u7ee7\u7eed\u3002',
        (u'en', u'fr', u'Correct!'): u'Correct!',
        (u'en', u'pt', u'Correct!'): u'Correto!',
        (u'en', u'es', u'hello world'): u'Hola Mundo',
        (u'en', u'pt', u'Please continue.'): u'Por favor continue.',
        (u'en', u'fr', u'hello world'): u'Bonjour le monde'
    }

     def __init__(self):
         """Initializes the emulator with no expected responses."""
         self.expected_responses = self.PREGENERATED_TRANSLATIONS
```

SarahZLipeles replied:
Thanks for the suggestion! Done.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,107 @@
+        """Returns the saved expected response for a given input. If no
+        response exists for the given input, returns a default response.
+
+        For more information on ISO 639-1 see:
```
Were you asked to place these here? I think it makes more sense to move this into the class/module docstring and give a one-sentence explanation for how it relates to the services: "this class uses ISO 639-1 compliant language codes to specify languages. To learn more about ISO 639-1, see: [link]".

SarahZLipeles replied:
Thanks for the suggestion! Fixed.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,107 @@
+
+        Args:
+            text: str. The text to be translated.
+            source_language_code: str. A allowlisted ISO 639-1 language code.
+            target_language_code: str. A allowlisted ISO 639-1 language code.
```
nit
```suggestion
            source_language_code: str. An allowlisted ISO 639-1 language code.
            target_language_code: str. An allowlisted ISO 639-1 language code.
```

SarahZLipeles replied:
Thanks for catching this. Fixed.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,107 @@
+            str. The translated text.
+        """
+        if not utils.is_valid_language_code(source_language_code):
+            raise ValueError('Invalid language code: %s' % source_language_code)
```
Please also add source/target to these error messages so it's obvious which one was wrong when the exception is raised.

SarahZLipeles replied:
Done.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,107 @@
+            https://www.w3schools.com/tags/ref_language_codes.asp
+
+        Args:
+            source_language_code: str. A allowlisted ISO 639-1 language code.
+            target_language_code: str. A allowlisted ISO 639-1 language code.
```
Ditto
```suggestion
            source_language_code: str. An allowlisted ISO 639-1 language code.
            target_language_code: str. An allowlisted ISO 639-1 language code.
```

SarahZLipeles replied:
Done.

brianrodri commented on core/platform/cloud_translate/cloud_translate_emulator_test.py:
```diff
@@ -0,0 +1,55 @@
+        self.emulator = cloud_translate_emulator.TranslateEmulator()
+
+    def test_init_prepopulates_responses(self):
+        self.assertDictEqual(
```
Please just use assertEqual, it'll choose the right thing.

SarahZLipeles replied:
Fixed.

brianrodri commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,56 @@
+        str. The translated text.
+    """
+    if source_language not in LANGUAGE_CODE_ALLOWLIST:
+        raise ValueError('Invalid language code: %s' % source_language)
```
Ditto, please make the errors more obvious.

SarahZLipeles replied:
Fixed.

brianrodri reviewed (changes requested):
Took another full pass, PTAL

Commit:
fix misnamed class in test

Commit:
move class constants to class body
update init docstring

Commit:
update invalid language code error messages to specify source or target
move ISO details to class docstring

sagangwee commented on core/platform/cloud_translate/cloud_translate_emulator.py:
```diff
@@ -0,0 +1,110 @@
+    API. This emulator can be used in backend testing, or a local dev
+    environment without access to the Cloud Translate API. Expected responses
+    must be passed in before using this emulator for testing. See
+    PREGENERATED_TRANSLATIONS above for some prepopulated responses.
```
Should this read "PREGENERATED_TRANSLATIONS below..."?

SarahZLipeles replied:
Yes, thanks for catching this! Fixed.

sagangwee commented on core/platform/cloud_translate/cloud_translate_services.py:
```diff
@@ -0,0 +1,57 @@
+        target_language: str. An allowlisted language code.
+
+    Returns:
+        str. The translated text.
```
Add possible exceptions here as well.

SarahZLipeles replied:
Done.

sagangwee commented on core/platform/cloud_translate/dev_mode_cloud_translate_services.py:
```diff
@@ -0,0 +1,49 @@
+        text: str. The text to be translated.
+        source_language: str. An allowlisted ISO 639-1 language code.
+        target_language: str. An allowlisted ISO 639-1 language code.
+
```
Add exceptions.

SarahZLipeles replied:
Done.

sagangwee reviewed (changes requested):
@TheoLipeles Thanks took a pass.

Commit:
switch assertEqual for assertDictEqual

Commit:
fix comments
add exceptions to comments

SarahZLipeles commented:
@brianrodri @sagangwee I responded to your comments, PTAL.

SarahZLipeles commented:
@sagangwee In reference to this comment: https://github.com/oppia/oppia/pull/12065#discussion_r605850025
Is it better to add a linter ignore here than to leave these checks in? I added them for the sole purpose of preventing the linting error, but I was thinking that they also help with debugging the tests themselves. For example, if one of these asserts were to fail, it would imply a problem with the usage of the mock api call rather than with `cloud_translate_services`.
Edit:
I also based this partially on how other tests are written, for example, in `cloud_taskqueue_services_test`, `mock_create_task` uses assertEqual for each of its parameters.

brianrodri reviewed (approved):
LGTM!

Commit:
add tests for add_expected_response
add instructions to dev_mode_cloud_translate_services

sagangwee commented:
Re: https://github.com/oppia/oppia/pull/12065#issuecomment-816325708, copying some thoughts from offline chat:

So one of the main reasons those assertions give me pause is generally if you're reading a test, you expect all your asserts to be at the end—it's a bit strange to me that it would be inside the mock function.

You might sometimes want to create helper methods for asserts when 'verifying a single fact about an object requires tricky logic or multiple asserts', but usually you would want to assert an explicit fact and the method's name would reflect that.

test_utils.swap_with_checks() with expected args is pretty neat and seems to fit the bill perfectly here. I think we can safely ignore the linter here as well as it is a mock function after all and we shouldn't feel forced to use the arguments.

Commit:
remove unnecessary checks

sagangwee commented on core/platform/cloud_translate/cloud_translate_services_test.py:
```diff
@@ -0,0 +1,67 @@
+                'translatedText': 'hola mundo'
+            }
+
+        with self.swap(
```
Can we use `swap_to_always_return` instead here?

SarahZLipeles replied:
Done. This is much cleaner!

sagangwee commented on core/platform/cloud_translate/cloud_translate_services_test.py:
```diff
@@ -0,0 +1,67 @@
+                text, source_language, target_language): # pylint: disable=unused-argument
+            self.fail(msg='mock api called')
+
+        with self.swap(
```
Can we use `swap_to_always_raises` instead here?

SarahZLipeles replied:
Done.

sagangwee reviewed (changes requested):
@TheoLipeles Took at the latest changes!

Commit:
replace mock api with swap_to_always_return and swap_to_always_raise

sagangwee reviewed (approved):
@TheoLipeles LGTM. Thanks!
